db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/TriggerDescriptorV3.java

```diff
@@ -0,0 +1,184 @@
+            TableDescriptor td,
+            UUID whenSPSId,
+            List<UUID> actionSPSIdList,
+            Timestamp creationTimestamp,
+            int[] referencedCols,
+            int[] referencedColsInTriggerAction,
+            List<String> triggerDefinitionList,
+            boolean referencingOld,
+            boolean referencingNew,
+            String oldReferencingName,
+            String newReferencingName,
+            String whenClauseText) {
+        super(
+            dataDictionary,
+            sd,
+            id,
+            name,
+            eventMask,
+            isBefore,
+            isRow,
+            isEnabled,
+            td,
+            whenSPSId,
+            actionSPSIdList,
+            creationTimestamp,
+            referencedCols,
+            referencedColsInTriggerAction,
+            triggerDefinitionList,
+            referencingOld,
+            referencingNew,
+            oldReferencingName,
+            newReferencingName,
+            whenClauseText);
+        this.version = 3;
+    }
+
+    //////////////////////////////////////////////////////////////
+    //
+    // FORMATABLE
+    //
+    //////////////////////////////////////////////////////////////
+    /**
+     * WARNING - If adding new fields to readExternal/writeExternal
+     * you must create a new subclass, and add the fields there.
+     * Otherwise old SPSDescriptors stored in sys.sysstatements may
+     * fail to deserialize properly.
+     *
+     * @param in read this.
+     * @throws IOException            thrown on error
+     * @throws ClassNotFoundException thrown on error
+     */
+    /**
+     * Read this object from a stream of stored objects.
+     *
+     * @param in read this.
+     * @throws IOException            thrown on error
+     * @throws ClassNotFoundException thrown on error
+     */
+    @Override
+    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+        super.readExternal(in);
+        version = in.readInt();
+
+        Object obj = in.readObject();
+        if (obj instanceof List) {
+            assert triggerDefinitionList == null : "triggerDefinition and triggerDefinitionList cannot both be defined";
+            triggerDefinitionList = (List<String>) obj;
+        }
+
+        obj = in.readObject();
+        if (obj instanceof List) {
+            assert actionSPSIdList == null : "actionSPSIdList and actionSPSId cannot both be defined";
+            actionSPSIdList= (List<UUID>) obj;
+        }
+
+    }
+
+    /**
+     * Write this object to a stream of stored objects.
+     *
+     * @param out write bytes here.
+     */
+    @Override
+    public void writeExternal(ObjectOutput out) throws IOException {
+        super.writeExternal(out);
+        out.writeInt(version);
+        out.writeObject(triggerDefinitionList.size() > 1 ? triggerDefinitionList : null);
+        out.writeObject(actionSPSIdList.size() > 1 ? actionSPSIdList : null);
+    }
```

Comment:
I have some concern about having null at the end of the write stream. How do we guarantee that we won't be reading fields from another object when calling readExternal() if these two fields are null ?

Reply:
The original concern was that we cannot differentiate the two scenarios 1) TriggerDescriptorV3 with null for triggerDefinitionList and actionSPSIdList followed by O1 starting with an List object, or 2) TriggerDescriptorV3 with non-null triggerDefinitionList and actionSPSIdList followed by O1. Based on a discussion with Arnaud, this is is a no issue, so closing this comment.